### ISSUE TITLE
Drop nogil?

### ISSUE THREAD

**Issue by bdice** [glotzerlab/freud#397]
After reviewing the Cython documentation, we may be using `nogil` in places where dropping the GIL isn't necessary. Dropping the GIL _may_ increase performance. We should test the correctness and performance of removing all `nogil` definitions. In particular, it's not quite clear how having/dropping the GIL affects our ability to raise Python exceptions from C++ code.

https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html

**Comment by vyasr** [comment]
Testing on my Mac, here are the results of `benchmarks/benchmark_density_RDF.py` with the current version of the code:
```
freud.density.RDF:
        rmax = 10.0, dr = 1.0, rmin = 0
      1000:    8.773 ms | 8772.501 ns per item
     10000:  545.855 ms | 54585.504 ns per item
Threads          1000          |         10000          |
      1   20.266 ms      1.00x | 1276.213 ms      1.00x |
      2   11.087 ms      1.83x |  673.562 ms      1.89x |
      3   10.095 ms      2.01x |  609.670 ms      2.09x |
      4    8.937 ms      2.27x |  546.529 ms      2.34x |

 ----------------
```
and here are the corresponding benchmarks after removing the `with nogil` statement in `density.pyx:RDF.accumulate` and the `nogil` modifier to `_density.pyx:RDF::accumulate`.
```
freud.density.RDF:
        rmax = 10.0, dr = 1.0, rmin = 0
      1000:    8.937 ms | 8937.082 ns per item
     10000:  551.799 ms | 55179.868 ns per item
Threads          1000          |         10000          |
      1   20.161 ms      1.00x | 1278.559 ms      1.00x |
      2   11.532 ms      1.75x |  681.465 ms      1.88x |
      3    9.880 ms      2.04x |  603.261 ms      2.12x |
      4    9.246 ms      2.18x |  548.136 ms      2.33x |

 ----------------
```

Also, the unit tests pass. This suggests that we're correct about not having to release the GIL. I'm running the same benchmarks on a Linux box with more cores to confirm that this behavior stands on something with enough processing power to properly test 4 threads.

**Comment by vyasr** [comment]
Here are the results on the Linux box.

Current code:
```
freud.density.RDF:
        rmax = 10.0, dr = 1.0, rmin = 0
      1000:    6.398 ms | 6398.461 ns per item
     10000:  108.409 ms | 10840.852 ns per item
Threads          1000          |         10000          |
      1   17.910 ms      1.00x | 1300.926 ms      1.00x |
      2   12.674 ms      1.41x |  679.127 ms      1.92x |
      3    9.213 ms      1.94x |  464.037 ms      2.80x |
      4    7.864 ms      2.28x |  379.704 ms      3.43x |
      5    6.589 ms      2.72x |  299.389 ms      4.35x |
      6    6.252 ms      2.86x |  276.862 ms      4.70x |
      7    5.765 ms      3.11x |  263.062 ms      4.95x |
      8    5.280 ms      3.39x |  223.320 ms      5.83x |
      9    4.984 ms      3.59x |  213.505 ms      6.09x |
     10    4.938 ms      3.63x |  197.624 ms      6.58x |
     11    4.861 ms      3.68x |  191.306 ms      6.80x |
     12    4.766 ms      3.76x |  189.753 ms      6.86x |
     13    4.663 ms      3.84x |  175.858 ms      7.40x |
     14    4.594 ms      3.90x |  167.234 ms      7.78x |
     15    4.561 ms      3.93x |  162.063 ms      8.03x |
     16    4.483 ms      4.00x |  152.968 ms      8.50x |
     17    4.449 ms      4.03x |  154.043 ms      8.45x |
     18    4.445 ms      4.03x |  154.193 ms      8.44x |
     19    4.412 ms      4.06x |  148.969 ms      8.73x |
     20    4.401 ms      4.07x |  135.264 ms      9.62x |
     21    4.409 ms      4.06x |  139.782 ms      9.31x |
     22    4.403 ms      4.07x |  130.934 ms      9.94x |
     23    4.437 ms      4.04x |  128.170 ms     10.15x |
     24    4.490 ms      3.99x |  125.075 ms     10.40x |
     25    4.552 ms      3.93x |  124.680 ms     10.43x |
```

Version without releasing the GIL:
```
freud.density.RDF:
        rmax = 10.0, dr = 1.0, rmin = 0
      1000:    6.755 ms | 6755.034 ns per item
     10000:  106.627 ms | 10662.676 ns per item
Threads          1000          |         10000          |
      1   17.936 ms      1.00x | 1306.516 ms      1.00x |
      2   12.435 ms      1.44x |  675.314 ms      1.93x |
      3    9.151 ms      1.96x |  463.715 ms      2.82x |
      4    8.143 ms      2.20x |  366.803 ms      3.56x |
      5    7.094 ms      2.53x |  335.093 ms      3.90x |
      6    6.610 ms      2.71x |  310.266 ms      4.21x |
      7    5.912 ms      3.03x |  272.308 ms      4.80x |
      8    5.450 ms      3.29x |  234.130 ms      5.58x |
      9    5.152 ms      3.48x |  220.606 ms      5.92x |
     10    4.929 ms      3.64x |  205.661 ms      6.35x |
     11    4.768 ms      3.76x |  177.728 ms      7.35x |
     12    4.702 ms      3.81x |  177.795 ms      7.35x |
     13    4.572 ms      3.92x |  178.695 ms      7.31x |
     14    4.564 ms      3.93x |  169.049 ms      7.73x |
     15    4.456 ms      4.02x |  168.949 ms      7.73x |
     16    4.477 ms      4.01x |  165.478 ms      7.90x |
     17    4.461 ms      4.02x |  166.502 ms      7.85x |
     18    4.401 ms      4.08x |  150.310 ms      8.69x |
     19    4.404 ms      4.07x |  145.783 ms      8.96x |
     20    4.368 ms      4.11x |  138.686 ms      9.42x |
     21    4.371 ms      4.10x |  141.531 ms      9.23x |
     22    4.435 ms      4.04x |  138.048 ms      9.46x |
     23    4.415 ms      4.06x |  125.566 ms     10.41x |
     24    4.387 ms      4.09x |  130.326 ms     10.02x |
     25    4.469 ms      4.01x |  125.817 ms     10.38x |
```

This seems like pretty conclusive evidence that we can safely perform our own threading without releasing the GIL since we're never touching Python objects in our C++ code.

**Comment by bdice** [comment]
Resolved in #411.
